providers/azure/azure.go

```diff
@@ -0,0 +1,298 @@
+	providers.RegisterProvider("azure", &ZoneDNSProvider{})
+}
+
+// Initialize Azure DNS provider.
+// Return nil in case of success otherwise an error
+func (zd *ZoneDNSProvider) Init(rootDomainName string) error {
+	zd.rootDomainName = strings.TrimRight(rootDomainName, ".")
+	zd.resourceGroup = os.Getenv("AZURE_RESOURCE_GROUP")
+	c := map[string]string{
+		"AZURE_CLIENT_ID":       os.Getenv("AZURE_CLIENT_ID"),
+		"AZURE_CLIENT_SECRET":   os.Getenv("AZURE_CLIENT_SECRET"),
+		"AZURE_SUBSCRIPTION_ID": os.Getenv("AZURE_SUBSCRIPTION_ID"),
+		"AZURE_TENANT_ID":       os.Getenv("AZURE_TENANT_ID")}
+
+	spt, err := helpers.NewServicePrincipalTokenFromCredentials(
+		c,
+		azure.PublicCloud.ResourceManagerEndpoint)
+	if err != nil {
+		return fmt.Errorf("error: %v", err)
+	}
+
+	zd.zoneClient = dns.NewZonesClient(c["AZURE_SUBSCRIPTION_ID"])
+	zd.zoneClient.Authorizer = autorest.NewBearerAuthorizer(spt)
+	zd.recordClient = dns.NewRecordSetsClient(c["AZURE_SUBSCRIPTION_ID"])
+	zd.recordClient.Authorizer = autorest.NewBearerAuthorizer(spt)
+
+	return nil
+}
+
+// Return the name of Azure DNS provider.
+func (zd *ZoneDNSProvider) GetName() string {
+	return "Azure Zone DNS"
+}
+
+// Check health of service, in this case fetch ZoneDNS info.
+// Returns nil in case of success otherwise an error.
+func (zd *ZoneDNSProvider) HealthCheck() error {
+	_, err := zd.zoneClient.Get(zd.resourceGroup, zd.rootDomainName)
+	return err
+}
+
+// Add a record to Azure DNS.
+func (zd *ZoneDNSProvider) AddRecord(record utils.DnsRecord) error {
+	return zd.changeRecord(record)
+}
+
+// Remove a record from Azure DNS.
+func (zd *ZoneDNSProvider) RemoveRecord(record utils.DnsRecord) error {
+	_, err := zd.recordClient.Delete(
+		zd.resourceGroup,
+		zd.rootDomainName,
+		strings.Replace(utils.UnFqdn(record.Fqdn), "." + zd.rootDomainName, "", -1),
+		dns.RecordType(record.Type),
+		"")
+
+	if err != nil {
+		return fmt.Errorf(
+			"failed to delete %s record named '%s' for DNS zone '%s': %v",
+			record.Type,
+			strings.TrimRight(record.Fqdn, "."),
+			zd.rootDomainName,
+			err,
+		)
+	}
+	return nil
+}
+
+// Update a record on Azure DNS.
+func (zd *ZoneDNSProvider) UpdateRecord(record utils.DnsRecord) error {
+	return zd.changeRecord(record)
+}
+
+// Fetch and return all records from Azure DNS.
+func (zd *ZoneDNSProvider) GetRecords() ([]utils.DnsRecord, error) {
+	dnsRecords := []utils.DnsRecord{}
+	err := zd.iterateRecords(func(recordSet dns.RecordSet) bool {
+		if recordSet.Name == nil || recordSet.Type == nil {
+			logrus.Error("Skipping invalid record set with nil name or type.")
+			return true
+		}
+		recordType := strings.TrimLeft(*recordSet.Type, "Microsoft.Network/dnszones/")
+		if !supportedRecordType(recordType) {
+			return true
+		}
+		name := formatAzureDNSName(*recordSet.Name, zd.rootDomainName)
+		targets := extractAzureTarget(&recordSet)
+		if targets == nil {
+			logrus.Errorf("Failed to extract target for '%s' with type '%s'.", name, recordType)
```

Comment:
is there a reason why it is logged in error? Empty (nil) records doesn't seem to qualify for the error